### PR TITLE
Add contains relationships

### PR DESCRIPTION
**Pull request by deads2k**
Adding a contains edge type and starting to split apart contained nodes to allow simpler edge creation and edge termination type guarantees.

@ncdc

### CODE REVIEW

**Comment by ncdc** [comment]
Initial review completed w/comments

**Comment by deads2k** [comment]
comments addressed.

**Comment by deads2k** [comment]
rebased.

**Comment by deads2k** [comment]
disallow multiple contains edges to a single node.

File: pkg/api/graph/graph.go [hunk 1]
```diff
@@ -153,6 +185,38 @@ func (g Graph) SuccessorEdges(node graph.Node, fn EdgeFunc, edgeKind ...string)
...
+		edge := g.EdgeBetween(n, node)
+		if len(allowedKinds) == 0 || allowedKinds.Has(g.EdgeKind(edge)) {
+			ret = append(ret, edge)
+		}
+	}
+
+	return ret
+}
+
+// InboundEdges returns all the outbound edges from node that are in the list of edgeKinds
```
Comment (ncdc) [inline pkg/api/graph/graph.go]: all the inbound edges to the node

Comment (deads2k) [inline pkg/api/graph/graph.go]: > all the inbound edges to the node

fixed

File: pkg/api/kubegraph/nodes/nodes_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		t.Fatalf("expected 1 edge, got %v", rcEdges)
+	}
+	if g.EdgeKind(rcEdges[0]) != osgraph.ContainsEdgeKind {
+		t.Errorf("expected %v, got %v", osgraph.ContainsEdgeKind, rcEdges[0])
+	}
+
+	uncastRCSpec := rcEdges[0].Tail()
+	rcSpec, ok := uncastRCSpec.(*ReplicationControllerSpecNode)
+	if !ok {
+		t.Fatalf("expected ptspec, got %v", uncastRCSpec)
```
Comment (ncdc) [inline pkg/api/kubegraph/nodes/nodes_test.go]: s/ptspec/rcspec/

File: pkg/api/kubegraph/nodes/types.go [hunk 3]
```diff
@@ -74,6 +108,64 @@ func (n ReplicationControllerNode) String() string {
...
+
+func ReplicationControllerSpecNodeName(o *kapi.ReplicationControllerSpec, ownerID osgraph.UniqueName) osgraph.UniqueName {
+	return osgraph.UniqueName(fmt.Sprintf("%s|%v", ReplicationControllerSpecNodeKind, ownerID))
+}
+
+type ReplicationControllerSpecNode struct {
+	osgraph.Node
+	*kapi.ReplicationControllerSpec
+
+	OwnerID osgraph.UniqueName
```
Comment (ncdc) [inline pkg/api/kubegraph/nodes/types.go]: OwnerName everywhere instead?

File: pkg/deploy/graph/nodes/nodes.go [hunk 4]
```diff
@@ -4,16 +4,23 @@ import (
...
 	osgraph "github.com/openshift/origin/pkg/api/graph"
+	kubegraph "github.com/openshift/origin/pkg/api/kubegraph/nodes"
 	depoyapi "github.com/openshift/origin/pkg/deploy/api"
 )
 
 // EnsureDeploymentConfigNode adds the provided deployment config to the graph if it does not exist
-func EnsureDeploymentConfigNode(g osgraph.MutableUniqueGraph, config *depoyapi.DeploymentConfig) *DeploymentConfigNode {
-	return osgraph.EnsureUnique(
+func EnsureDeploymentConfigNode(g osgraph.MutableUniqueGraph, dc *depoyapi.DeploymentConfig) *DeploymentConfigNode {
+	dcID := DeploymentConfigNodeName(dc)
```
Comment (ncdc) [inline pkg/deploy/graph/nodes/nodes.go]: dcNodeName

Comment (deads2k) [inline pkg/deploy/graph/nodes/nodes.go]: weird, thought I got those

**Comment by deads2k** [comment]
comments addressed

**Comment by deads2k** [comment]
one a day?

**Comment by ncdc** [comment]
Other than lots of funcs missing Godoc, LGTM

**Comment by deads2k** [comment]
godoc added for non-boilerplate. [merge]

**Pull request merged**